### PR TITLE
[SPARK-21383][Core] Fix the YarnAllocator allocates more Resource

### PR DESCRIPTION
**Pull request by libratiger**
## What changes were proposed in this pull request?
When NodeManagers launching Executors,
the `missing` value will exceed the
real value when the launch is slow, this can lead to YARN allocates more resource.

We add the `numExecutorsRunning` when calculate the `missing` to avoid this.

## How was this patch tested?
Test by experiment.

### CODE REVIEW

**Comment by vanzin** [comment]
ok to test

**Comment by vanzin** [comment]
the missing value will excel the

Not sure what that means. Maybe "exceed"?

File: resource-managers/yarn/src/main/scala/org/apache/spark/deploy/yarn/YarnAllocator.scala [hunk 1]
```diff
@@ -505,32 +508,37 @@ private[yarn] class YarnAllocator(
...
-                ).run()
-                updateInternalState()
-              } catch {
-                case NonFatal(e) =>
-                  logError(s"Failed to launch executor $executorId on container $containerId", e)
-                  // Assigned container should be released immediately to avoid unnecessary resource
-                  // occupation.
-                  amClient.releaseAssignedContainer(containerId)
+          try {
+            numExecutorToBeLaunched += 1
```
Comment (vanzin) [inline resource-managers/yarn/src/main/scala/org/apache/spark/deploy/yarn/YarnAllocator.scala]: This is not thread safe. `+=` is two operations.

File: resource-managers/yarn/src/main/scala/org/apache/spark/deploy/yarn/YarnAllocator.scala [hunk 2]
```diff
@@ -505,32 +508,37 @@ private[yarn] class YarnAllocator(
...
+                    // Assigned container should be released immediately
+                    // to avoid unnecessary resource occupation.
+                    amClient.releaseAssignedContainer(containerId)
+                }
               }
-            }
-          })
+            })
+          } finally {
+            numExecutorToBeLaunched -= 1
```
Comment (vanzin) [inline resource-managers/yarn/src/main/scala/org/apache/spark/deploy/yarn/YarnAllocator.scala]: Same here.

File: resource-managers/yarn/src/main/scala/org/apache/spark/deploy/yarn/YarnAllocator.scala [hunk 3]
```diff
@@ -82,6 +82,8 @@ private[yarn] class YarnAllocator(
 
   @volatile private var numExecutorsRunning = 0
 
+  @volatile private var numExecutorToBeLaunched = 0
```
Comment (vanzin) [inline resource-managers/yarn/src/main/scala/org/apache/spark/deploy/yarn/YarnAllocator.scala]: A better name would be `numExecutorsStarting`.

File: resource-managers/yarn/src/main/scala/org/apache/spark/deploy/yarn/YarnAllocator.scala (hunk 2 above)
Comment (vanzin) [inline resource-managers/yarn/src/main/scala/org/apache/spark/deploy/yarn/YarnAllocator.scala]: Also, shouldn't this be done in the actual thread that launching the executor? Otherwise you're decrementing the counter as soon as the task is submitted for execution, but you're not really waiting for the task to execute.

File: resource-managers/yarn/src/main/scala/org/apache/spark/deploy/yarn/YarnAllocator.scala (hunk 3 above)
Comment (djvulee) [inline resource-managers/yarn/src/main/scala/org/apache/spark/deploy/yarn/YarnAllocator.scala]: OK，I will change the name.

File: resource-managers/yarn/src/main/scala/org/apache/spark/deploy/yarn/YarnAllocator.scala (hunk 2 above)
Comment (djvulee) [inline resource-managers/yarn/src/main/scala/org/apache/spark/deploy/yarn/YarnAllocator.scala]: Yes, you're right. When I test the code by experiment, i decrease the `numExecutorToBeLaunched` in the `updateInternalState` function, but I later found this may impact the test.

I will fix this soon.

**Comment by djvulee** [comment]
I just update the code, and test by experiment, can you take a look at @vanzin

File: resource-managers/yarn/src/main/scala/org/apache/spark/deploy/yarn/YarnAllocator.scala [hunk 4]
```diff
@@ -270,7 +272,7 @@ private[yarn] class YarnAllocator(
       logDebug("Allocated containers: %d. Current executor count: %d. Cluster resources: %s."
         .format(
           allocatedContainers.size,
-          numExecutorsRunning,
+          numExecutorsRunning.get,
```
Comment (tgravescs) [inline resource-managers/yarn/src/main/scala/org/apache/spark/deploy/yarn/YarnAllocator.scala]: it woudl be nice to print the numExecutorsStarting here as well.

File: resource-managers/yarn/src/main/scala/org/apache/spark/deploy/yarn/YarnAllocator.scala [hunk 5]
```diff
@@ -294,7 +296,8 @@ private[yarn] class YarnAllocator(
   def updateResourceRequests(): Unit = {
     val pendingAllocate = getPendingAllocate
     val numPendingAllocate = pendingAllocate.size
-    val missing = targetNumExecutors - numPendingAllocate - numExecutorsRunning
+    val missing = targetNumExecutors - numPendingAllocate -
+      numExecutorsStarting.get - numExecutorsRunning.get
```
Comment (tgravescs) [inline resource-managers/yarn/src/main/scala/org/apache/spark/deploy/yarn/YarnAllocator.scala]: can you also add in a debug message here, something like below.  I found this very useful when debugging this issue and think it would be useful for debugging other allocating issues in the future.

logDebug(s"Updating resource requests, target: $targetNumExecutors, pending: " +
 +      s"$numPendingAllocate, running: $numExecutorsRunning, executorsStarting $numExecutorsPendingStart")

Comment (djvulee) [inline resource-managers/yarn/src/main/scala/org/apache/spark/deploy/yarn/YarnAllocator.scala]: Thanks for your advice! I just add the debug info.

File: resource-managers/yarn/src/main/scala/org/apache/spark/deploy/yarn/YarnAllocator.scala [hunk 6]
```diff
@@ -242,7 +244,7 @@ private[yarn] class YarnAllocator(
     if (executorIdToContainer.contains(executorId)) {
       val container = executorIdToContainer.get(executorId).get
       internalReleaseContainer(container)
-      numExecutorsRunning -= 1
```
Comment (vanzin) [inline resource-managers/yarn/src/main/scala/org/apache/spark/deploy/yarn/YarnAllocator.scala]: This doesn't need to be an atomic integer because this method is synchronized already.

File: resource-managers/yarn/src/main/scala/org/apache/spark/deploy/yarn/YarnAllocator.scala [hunk 7]
```diff
@@ -525,8 +535,9 @@ private[yarn] class YarnAllocator(
               } catch {
                 case NonFatal(e) =>
                   logError(s"Failed to launch executor $executorId on container $containerId", e)
-                  // Assigned container should be released immediately to avoid unnecessary resource
-                  // occupation.
+                  // Assigned container should be released immediately
+                  // to avoid unnecessary resource occupation.
+                  numExecutorsStarting.decrementAndGet()
```
Comment (vanzin) [inline resource-managers/yarn/src/main/scala/org/apache/spark/deploy/yarn/YarnAllocator.scala]: Safer to put this in a `finally`, no?

**Review by vanzin — COMMENTED** [review]
The changes to `numExecutorsRunning` are not necessary, but at the same time they don't hurt. I'll let Tom take a look since he left feedback earlier.

File: resource-managers/yarn/src/main/scala/org/apache/spark/deploy/yarn/YarnAllocator.scala (hunk 6 above)
Comment (djvulee) [inline resource-managers/yarn/src/main/scala/org/apache/spark/deploy/yarn/YarnAllocator.scala]: Yes, I just try to keep consistency with `numExecutorsStarting`

File: resource-managers/yarn/src/main/scala/org/apache/spark/deploy/yarn/YarnAllocator.scala (hunk 7 above)
Comment (djvulee) [inline resource-managers/yarn/src/main/scala/org/apache/spark/deploy/yarn/YarnAllocator.scala]: Yes, it is more robust. I have update the code.

File: resource-managers/yarn/src/main/scala/org/apache/spark/deploy/yarn/YarnAllocator.scala [hunk 8]
```diff
@@ -525,9 +534,11 @@ private[yarn] class YarnAllocator(
               } catch {
                 case NonFatal(e) =>
                   logError(s"Failed to launch executor $executorId on container $containerId", e)
-                  // Assigned container should be released immediately to avoid unnecessary resource
-                  // occupation.
+                  // Assigned container should be released immediately
+                  // to avoid unnecessary resource occupation.
                   amClient.releaseAssignedContainer(containerId)
+              } finally {
+                numExecutorsStarting.decrementAndGet()
```
Comment (tgravescs) [inline resource-managers/yarn/src/main/scala/org/apache/spark/deploy/yarn/YarnAllocator.scala]: why did you move this out of updateInternalState?  I would rather see those stay together because updateInternalState is synchronized and if they aren't changed at the same time you could end up with the wrong numbers. numExecutorsStarting could still be say 1 when you have already added that to the # running and then updateResourceRequests gets an incorrect number.

Comment (vanzin) [inline resource-managers/yarn/src/main/scala/org/apache/spark/deploy/yarn/YarnAllocator.scala]: If you move this to `updateInternalState`, then if there's a problem in `ExecutorRunnable.run` this counter will not be decremented.

Comment (tgravescs) [inline resource-managers/yarn/src/main/scala/org/apache/spark/deploy/yarn/YarnAllocator.scala]: sorry don't follow. updateInternalState is called right after the run if it succeeds.  You do still need to handle the failure as well but in the failure the running is never incremented

Comment (vanzin) [inline resource-managers/yarn/src/main/scala/org/apache/spark/deploy/yarn/YarnAllocator.scala]: this is "starting" not "running".

Comment (tgravescs) [inline resource-managers/yarn/src/main/scala/org/apache/spark/deploy/yarn/YarnAllocator.scala]: Note I thought the last version handled this properly where he had the decrement in the catch block

File: resource-managers/yarn/src/main/scala/org/apache/spark/deploy/yarn/YarnAllocator.scala [hunk 9]
```diff
@@ -503,8 +511,9 @@ private[yarn] class YarnAllocator(
         allocatedContainerToHostMap.put(containerId, executorHostname)
       }
 
-      if (numExecutorsRunning < targetNumExecutors) {
+      if (numExecutorsRunning.get < targetNumExecutors) {
         if (launchContainers) {
+          numExecutorsStarting.incrementAndGet()
```
Comment (tgravescs) [inline resource-managers/yarn/src/main/scala/org/apache/spark/deploy/yarn/YarnAllocator.scala]: note if we change the below back this needs to move outside if(launchContainers)

File: resource-managers/yarn/src/main/scala/org/apache/spark/deploy/yarn/YarnAllocator.scala (hunk 8 above)
Comment (tgravescs) [inline resource-managers/yarn/src/main/scala/org/apache/spark/deploy/yarn/YarnAllocator.scala]: not sure I follow your reference between starting and running.  I was just saying in the failure case it doesn't matter because you aren't going to overcount.

If we don't decrement the starting inside of updateInternalState we have the possibility to over count because running and starting aren't incremented/decremented within the synchronize block.  We don't want to do that.

Comment (vanzin) [inline resource-managers/yarn/src/main/scala/org/apache/spark/deploy/yarn/YarnAllocator.scala]: `numExecutorsStarting` is incremented before the task is submitted and needs to be decremented after the task finishes, successfully or not. That's exactly what the current code does.

Comment (tgravescs) [inline resource-managers/yarn/src/main/scala/org/apache/spark/deploy/yarn/YarnAllocator.scala]: yes but its a bug right now as the numbers can be wrong. Are you looking at the synchronization?

Right now everything is called synchronized up to the point of launcher pool to do the ExecutorRunnable.  At this point running is not incremented, pending is decremented and we now increment Starting.  That is fine.

But when the ExecutorRunnable finishes the only place its called synchronized is in updateInternalState.  This right now increments running but does not decrement starting.  if updateResourceRequests gets called (which is synchronized), Right after updateInternalState (which leave the syncrhonized) but before the finally block executes and decrements starting the total number can be more then it really is.  That executor is counted as both running and starting

Comment (vanzin) [inline resource-managers/yarn/src/main/scala/org/apache/spark/deploy/yarn/YarnAllocator.scala]: Ok, I see. The current code can double count the same executor as starting and running, while the previous code could count it as starting even though it failed to start (for a really small window), but that is a self-healing situation while the previous can have some adverse effects.

Comment (tgravescs) [inline resource-managers/yarn/src/main/scala/org/apache/spark/deploy/yarn/YarnAllocator.scala]: yep, the fact that its still marked as Starting even though failed will fix itself next loop through. Its no different then if we didn't know it failed and it was still in the ExecutorRunnable.run code.

Comment (djvulee) [inline resource-managers/yarn/src/main/scala/org/apache/spark/deploy/yarn/YarnAllocator.scala]: I agree that put `numExecutorsStarting.decrementAndGet()` together with `numExecutorsRunning.incrementAndGet()` in the `updateInternalState` is better if we can.

Why I try to put `numExecutorsStarting.decrementAndGet()` in the `finally` block is that if there some Exceptions is not `NonFatal`, and caught by the following code, we may can not allocated resources as we  specified, this is the same as @vanzin worried.

We may double the count in the current code, but this only slow down the allocation rate for a small time.

Comment (tgravescs) [inline resource-managers/yarn/src/main/scala/org/apache/spark/deploy/yarn/YarnAllocator.scala]: If its a fatal error I don't really expect us to continue so dynamic allocation doesn't matter. If you know a case where we would recover from fatal exception, I'm fine with adding in another catch there to decrement in the fatal case as well.

**Comment by djvulee** [comment]
I update the code, please take a look at @vanzin @tgravescs

**Comment by tgravescs** [comment]
LGTM.  @vanzin  anything further?

**Comment by vanzin** [comment]
Merging to master / 2.2.
